sqlccl: resume backups and restores when coordinator fails

benesch opened the pull request:
~NOT READY TO MERGE~

~These commits desperately need to be resplit, and I need to figure out how to test this, but the resumability pieces are all in place. :tada:~

~Don't bother reviewing for anything but high-level feedback yet!~

This is ready for serious review—and serious apologies in advance about the size of this PR. I've done my best to split it into bite-sized commits, and tagged each below with the team it needs review from.

* sql/jobs: add FinishedWith convenience method — @cockroachdb/sql-async
* sql/jobs: expose per-server information needed by jobs — @cockroachdb/sql-async
* sql/jobs: add storage.NodeLiveness mock— @cockroachdb/core
* sql/jobs: manage job leases and resumability — @cockroachdb/core, @cockroachdb/sql-async
* sqlccl: further split restore setup and execution — @cockroachdb/sql-ccl
* sqlccl: resume backups when coordinator fails— @cockroachdb/sql-ccl
* sqlccl: resume restores when coordinator fail— @cockroachdb/sql-ccl
* sqlccl: test resuming backups and restores— @cockroachdb/sql-ccl

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17301)
<!-- Reviewable:end -->

benesch committed:
sql/jobs: manage job leases and resumability

Teach jobs.Registry to resume jobs when their node liveness-based lease
expires. See the backup/restore RFC addendum (e103024, #16633) for
details.

benesch committed:
sql/jobs: add FinishedWith convenience method

Currently, all three types of long-running jobs (backups, restores, and
schema changes) all use the same pattern to report status after the job
has completed:

    err := job(...)
    if err != nil {
        job.Failed(ctx, err)
        return err
    }
    if err := job.Succeeded(ctx, err); err != nil {
        log.Errorf(ctx, "ignoring job.Succeeded error: %s", err)
    }

Replace this pattern with a new Job method, FinishedWith, that does
exactly that. This commit only switches backup and restore to use the
new method; switching schema changes over will require a bit of code
golf in another commit. The new pattern is:

    jobErr := job(...)
    if err := job.FinishedWith(ctx, err); err != nil {
        // This error is *always* important enough to return to the
        // caller.
        return err
    }
    if jobErr != nil {
        return jobErr
    }

Perhaps more importantly, this refactor makes reporting errors with the
forthcoming resumable coordinator much easier. Specifically, the error
returned by the job may not be a true failure but an error generated by
a cancelation request when e.g. the coordinator loses its lease. In that
case, we want to instead return to the user roachpb.AmbiguousError, not
e.g. context.Canceled. Intercepting errors like this was not possible
with the old pattern, but is possible with the new pattern: the
rewritten error is simply returned from FinishedWith.

benesch committed:
sqlccl: further split restore setup and execution

To prepare for resumability, 6d788d5 (#17232) divided restore's setup
logic from its execution logic. After 6d788d5, the output from the setup
phase was a map from old table IDs to new IDs. In theory, this map
should have been sufficient information to resume a restore.

Practice, of course, is never so simple. I missed that the setup logic
was also responsible for reassigning parent IDs. The output from the
setup phase needs to be a map from old table ID to (new table ID, new
parent ID).

This commit teaches the setup phase to return such a map, rather than
mutating the table descriptors in place to assign the new parent ID. The
execution phase assumes the responsibility of updating the table
descriptor with the dictated parent ID.

benesch committed:
sqlccl: resume backups when coordinator fails

benesch committed:
sqlccl: resume restores when coordinator fails

benesch committed:
sqlccl: test resuming backups and restores

benesch committed:
sql/jobs: add storage.NodeLiveness mock

benesch committed:
sql/jobs: expose per-server information needed by jobs

Backup needs the node ID, the cluster ID, and a *gossip.Gossip  on
resume.

danhhz commented:
bravo on getting this working! some initial thoughts while i dig in deeper

---

Review status: 0 of 17 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/backup.go, line 442 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO0zp2TlTgeghB4LJH:-KqO0zp2TlTgeghB4LJI:b-ywic5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/ccl/sqlccl/backup.go#L442)):*
> ```Go
> 		}
> 	} else if job.ID() != nil {
> 		// TODO(benesch): distinguish between a missing checkpoint, which simply
> ```

the caller to backup knows whether it's a resumption or not, no?

lets parse the checkpointed descriptor in resumeBackupHook and pass it into backup (so we can don't need this workaround)

---

*[pkg/ccl/utilccl/intervalccl/overlap_merge.go, line 15 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO1ot3ZQRu7DfQt3qg:-KqO1ot3ZQRu7DfQt3qh:b-xiyl3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/ccl/utilccl/intervalccl/overlap_merge.go#L15)):*
> ```Go
> 	"sort"
> 
> 	"github.com/cockroachdb/cockroach/pkg/roachpb"
> ```

oof, let's not introduce this dep

---

*[pkg/testutils/fake/id.go, line 17 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO255cqzS-e14hn3xG:-KqO255cqzS-e14hn3xH:b-nub1bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/testutils/fake/id.go#L17)):*
> ```Go
> // Author: Nikhil Benesch (nikhil.benesch@gmail.com)
> 
> package fake
> ```

i don't love this package. these sorts of things usually end up being too specific to the test that originally wanted them and also make very undesirable dependency graphs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqO2RmDQSuYQ5sy9SFA:b-5qoi4e)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup.go, line 442 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO0zp2TlTgeghB4LJH:-KqO8suW6AvDOGK2KfFu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/ccl/sqlccl/backup.go#L442)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

the caller to backup knows whether it's a resumption or not, no?

lets parse the checkpointed descriptor in resumeBackupHook and pass it into backup (so we can don't need this workaround)
</blockquote></details>

Done.

---

*[pkg/ccl/utilccl/intervalccl/overlap_merge.go, line 15 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO1ot3ZQRu7DfQt3qg:-KqOJE0QvXYzACYZZvdu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/ccl/utilccl/intervalccl/overlap_merge.go#L15)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

oof, let's not introduce this dep
</blockquote></details>

Done.

---

*[pkg/testutils/fake/id.go, line 17 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqO255cqzS-e14hn3xG:-KqOJF2NJImmgc9caPuf:bxu80z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/94276ebff8dc11048d0bb8959badce5657c7722e/pkg/testutils/fake/id.go#L17)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i don't love this package. these sorts of things usually end up being too specific to the test that originally wanted them and also make very undesirable dependency graphs
</blockquote></details>

Done. Moved to sql/jobs/helpers_test.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Reviewed 1 of 1 files at r11, 9 of 12 files at r12, 3 of 3 files at r13.
Review status: 8 of 15 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/jobs.go, line 107 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPbJMRXKV9WS8zX4oe:-KqPbJMRXKV9WS8zX4of:bs61ku6) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L107)):*
> ```Go
> // remembers the assigned ID of the job in the Job. The job information is read
> // from the Record field at the time Created is called. If cancelFn is not
> // WithoutCancel, the Registry will automatically acquire a lease for this job
> ```

nit: "If cancelFn is not nil" (what you have is not quite literally true)

---

*[pkg/sql/jobs/jobs.go, line 222 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPbc-LPP1ys1xsUqYi:-KqPbc-LPP1ys1xsUqYj:bb8jze7) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L222)):*
> ```Go
> 		return err
> 	}
> 	j.registry.unregister(*j.id)
> ```

should this be deferred, so it runs even if j.update errors?

---

*[pkg/sql/jobs/registry_external_test.go, line 93 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPcxwnwYFHetrxk5hY:-KqPcxwozemQ9ldJ8x1z:b4exje9) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/registry_external_test.go#L93)):*
> ```Go
> 		// call to nodeLiveness.GetLivenesses. Only after we've witnessed one call
> 		// for each job, plus one more call, can we be sure that all work has been
> 		// copmleted.
> ```

nit: completed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301)*
<!-- Sent from Reviewable.io -->

danhhz commented:
okay, looked at everything but the lease code. I have no substantive comments

---

Review status: 8 of 15 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 429 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPfjDQap4vg3MLqkq-:-KqPfjDQap4vg3MLqkq0:b9lhxva) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/backup.go#L429)):*
> ```Go
> 	}{}
> 
> 	var checkpointMu syncutil.Mutex
> ```

what is this used for?

---

*[pkg/ccl/sqlccl/backup.go, line 857 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPgdVk7Ju8dICAD_XV:-KqPgdVk7Ju8dICAD_XW:boza3q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/backup.go#L857)):*
> ```Go
> 	sql.AddPlanHook(backupPlanHook)
> 	sql.AddPlanHook(showBackupPlanHook)
> 	jobs.AddResumeHook(resumeBackupHook)
> ```

nit: I think backupResumeHook reads more clearly

---

*[pkg/ccl/sqlccl/backup_test.go, line 693 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPeinC_EJP5CQJXOkW:-KqPeinC_EJP5CQJXOkX:bmbvq8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5a2a59bb58706965b695913d70693ccaacac0a/pkg/ccl/sqlccl/backup_test.go#L693)):*
> ```Go
> // verifies that backup and restore do not re-perform work the checkpoint claims
> // to have completed.
> func TestBackupRestoreResume(t *testing.T) {
> ```

This should probably be two tests. (or at least two subtests)

---

*[pkg/ccl/sqlccl/progress.go, line 53 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPgzzAHl38nRvWX7n0:-KqPgzzAHl38nRvWX7n1:bsg0ctd) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/progress.go#L53)):*
> ```Go
> 	jpl.mu.completedChunks++
> 	fraction := float32(jpl.mu.completedChunks) / float32(jpl.totalChunks)
> 	fraction = fraction*(1-jpl.startFraction) + jpl.startFraction
> ```

nice! this is an elegant way of handling this

---

*[pkg/ccl/sqlccl/restore.go, line 387 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPhUFMLnNA3lR51Fv9:-KqPhUFNTfrVVBFTvUyP:btjabf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a38c5a1fa72cc657a7e85f2ef86e3268c58ef8c/pkg/ccl/sqlccl/restore.go#L387)):*
> ```Go
> ) ([]importEntry, hlc.Timestamp, error) {
> 	// Put the covering for the already-completed spans into the
> 	// OverlapCoveringMerge input first.
> ```

mention why this is important

---

*[pkg/ccl/sqlccl/restore.go, line 961 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPhhYHLspxRC4Li70c:-KqPhhYHLspxRC4Li70d:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a38c5a1fa72cc657a7e85f2ef86e3268c58ef8c/pkg/ccl/sqlccl/restore.go#L961)):*
> ```Go
> func init() {
> 	sql.AddPlanHook(restorePlanHook)
> 	jobs.AddResumeHook(resumeRestoreHook)
> ```

ditto

---

*[pkg/sql/jobs/registry.go, line 103 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPeu3mdByX6G0elYlg:-KqPeu3mdByX6G0elYlh:b-i27por) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5a2a59bb58706965b695913d70693ccaacac0a/pkg/sql/jobs/registry.go#L103)):*
> ```Go
> 
> // DefaultAdoptInterval is a reasonable interval at which to poll system.jobs
> // for jobs with expired leases. Mutable for testing.
> ```

Mutations only have an effect if made before server start, right? You should probably document that

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqPhlmrJE3KeeHlVDV5:b-t8yf6g)*
<!-- Sent from Reviewable.io -->

benesch commented:
Sweeet, TFTR!

---

Review status: 0 of 15 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup.go, line 429 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPfjDQap4vg3MLqkq-:-KqQ90I0XTJ_I8CC9U7k:btc0e81) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/backup.go#L429)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

what is this used for?
</blockquote></details>

Same as before; it ensures that two nodes don't attempt to write `BACKUP-CHECKPOINT` simultaneously without blocking import responses. 

---

*[pkg/ccl/sqlccl/backup.go, line 857 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPgdVk7Ju8dICAD_XV:-KqQ9MwEJSB3w66jamtc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/backup.go#L857)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: I think backupResumeHook reads more clearly
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/progress.go, line 53 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPgzzAHl38nRvWX7n0:-KqQA-TxO9jEmVrK29PA:b-snveef) ([raw file](https://github.com/cockroachdb/cockroach/blob/defb97a89868ef8ab806a9a49d2eaed5ec982f87/pkg/ccl/sqlccl/progress.go#L53)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nice! this is an elegant way of handling this
</blockquote></details>

The credit is all @jordanlewis's. https://github.com/cockroachdb/cockroach/commit/ca14df1c5d816c3c87d5b6e59d7b6f14c88b8adb

---

*[pkg/ccl/sqlccl/restore.go, line 387 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPhUFMLnNA3lR51Fv9:-KqQBIfHJHLVaEzeV9E9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a38c5a1fa72cc657a7e85f2ef86e3268c58ef8c/pkg/ccl/sqlccl/restore.go#L387)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

mention why this is important
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/restore.go, line 961 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPhhYHLspxRC4Li70c:-KqQANqXXmwMuAasFp0h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a38c5a1fa72cc657a7e85f2ef86e3268c58ef8c/pkg/ccl/sqlccl/restore.go#L961)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.go, line 107 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPbJMRXKV9WS8zX4oe:-KqPeTh-6jMZrGEYY3DQ:b-p9xrw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L107)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "If cancelFn is not nil" (what you have is not quite literally true)
</blockquote></details>

Oops, forgot I changed those semantics after writing the comment. Thanks!

---

*[pkg/sql/jobs/jobs.go, line 222 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPbc-LPP1ys1xsUqYi:-KqQBioUBIqyACVhMyqN:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L222)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

should this be deferred, so it runs even if j.update errors?
</blockquote></details>

Good call. Done.

---

*[pkg/sql/jobs/registry.go, line 103 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPeu3mdByX6G0elYlg:-KqQDBJNd1xiQZywQKPI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5a2a59bb58706965b695913d70693ccaacac0a/pkg/sql/jobs/registry.go#L103)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Mutations only have an effect if made before server start, right? You should probably document that
</blockquote></details>

Done.

---

*[pkg/sql/jobs/registry_external_test.go, line 93 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPcxwnwYFHetrxk5hY:-KqQC1PDvBTQ2wvuqJEB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/registry_external_test.go#L93)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: completed
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqQDDsmtLBFTEfaOiG_:b-xtobiz)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed up through 137973e7bd9f906653b0f0d80f58b7bc1ce5ea46, which you said was the important one. I basically just have one big question.

---

Reviewed 17 of 17 files at r9, 3 of 3 files at r10, 1 of 1 files at r11, 9 of 12 files at r12, 1 of 3 files at r13.
Review status: 0 of 15 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/jobs.go, line 254 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQDvJ9SmFhbaK6E8hO:-KqQDvJA0FR_Dv21OSgD:bqaltbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L254)):*
> ```Go
> 		// when using errors.Wrapf, and we'd much rather have false positives (too
> 		// many ambiguous results) than false negatives (too few ambiguous results).
> 		return roachpb.NewAmbiguousResultError("job lease expired")
> ```

Would there be any harm in including the error in this message?

---

*[pkg/sql/jobs/jobs.go, line 330 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEEbbPFpWTFJCU9fd:-KqQEEbbPFpWTFJCU9fe:b-u7q1lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L330)):*
> ```Go
> 		return err
> 	}
> 	// Don't need to lock because we're the only one who has a handle on this
> ```

Is there any reason not to lock though? I know you didn't write this, but you've now separated it farther from the creation of the payload. This doesn't look like a hot code path, and if this assumption is ever broken then getting cute with this won't have been it.

---

*[pkg/sql/jobs/jobs.go, line 444 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEw2ENEJblhytb18U:-KqQEw2ENEJblhytb18V:b7hxi8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L444)):*
> ```Go
> 	return j.update(ctx, StatusRunning, func(payload *Payload) (bool, error) {
> 		if !payload.Lease.Equal(oldLease) {
> 			return false, errors.New("existing lease did not match expected lease")
> ```

Mind including the leases in the error?

---

*[pkg/sql/jobs/jobs.go, line 455 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFC3_XBjTb1E-VUkA:-KqQFC3_XBjTb1E-VUkB:b-p21vfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L455)):*
> ```Go
> 
> func (j *Job) doCancel() {
> 	j.cancel()
> ```

`Job` has a `cancel` method and a `doCancel` method, neither of which takes any parameters? That seems likely to cause confusion.

---

*[pkg/sql/jobs/jobs.proto, line 33 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFKoCjgC8lQRHnVOE:-KqQFKoCjgC8lQRHnVOF:bso2qmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.proto#L33)):*
> ```protobuf
>   ];
>   // The epoch of the lease holder's node liveness entry.
>   int64 epoch = 2;
> ```

Piggy-backing on node liveness is nice from a reuse perspective, but it seems like the wrong model. Our epoch-based leases are designed to time out quickly, in order to minimize unavailability to transactional workloads. "Quickly" means 9 seconds  today, but if Spencer has his way it'll soon be 4.5 seconds (#16164). This means that anything that disrupts the coordinator node or the node liveness range's leaseholder for a couple seconds will cancel the job.

How expensive is it for a job to get canceled? If it's a fairly big deal, then I'd argue we need to be using longer leases for them to provide extra cushion.

For backward-compatible migrations, we're using a duration of 1 minute (https://github.com/cockroachdb/cockroach/blob/master/pkg/migration/sqlmigrations/migrations.go#L39), and our migrations take much less time than these jobs. The lease implementation that they use (https://github.com/cockroachdb/cockroach/blob/master/pkg/internal/client/lease.go) could be adapted for this if needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqQHS4JiwGMa-gSHC9h:b-wvofux)*
<!-- Sent from Reviewable.io -->

benesch commented:
To address your big question, it's no longer very expensive to cancel a
job. You'll redo at most a minute of BACKUP work and even less of RESTORE
work. Still far from ideal, though.

For context, the motivation for liveness-based leases was a (half-formed)
suggestion from @petermattis in
https://github.com/cockroachdb/cockroach/pull/16633#issuecomment-314456684.

In any case, switching to time-based leases means this won't make the code
freeze. Part of me thinks the right approach is to merge this with
livnesses-based leases, modulo addressing your feedback, with the
expectation that we'll switch to a time-based lease system like schema
changes or migrations use if we run into trouble. Or perhaps liveness-based
leases are fraught with enough peril this shouldn't land as is. I don't
have a good enough mental model of leasing to argue one way or another.

@danhhz or @petermattis, do you have thoughts?

On Mon, Jul 31, 2017 at 8:55 PM Alex Robinson <notifications@github.com>
wrote:

> Reviewed up through 137973e
> <https://github.com/cockroachdb/cockroach/commit/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46>,
> which you said was the important one. I basically just have one big
> question.
> ------------------------------
>
> Reviewed 17 of 17 files at r9, 3 of 3 files at r10, 1 of 1 files at r11, 9
> of 12 files at r12, 1 of 3 files at r13.
> Review status: 0 of 15 files reviewed at latest revision, 18 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/sql/jobs/jobs.go, line 254 at r12
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQDvJ9SmFhbaK6E8hO:-KqQDvJA0FR_Dv21OSgD:bqaltbp>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L254>):*
>
> 		// when using errors.Wrapf, and we'd much rather have false positives (too
> 		// many ambiguous results) than false negatives (too few ambiguous results).
> 		return roachpb.NewAmbiguousResultError("job lease expired")
>
> Would there be any harm in including the error in this message?
> ------------------------------
>
> *pkg/sql/jobs/jobs.go, line 330 at r12
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEEbbPFpWTFJCU9fd:-KqQEEbbPFpWTFJCU9fe:b-u7q1lm>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L330>):*
>
> 		return err
> 	}
> 	// Don't need to lock because we're the only one who has a handle on this
>
> Is there any reason not to lock though? I know you didn't write this, but
> you've now separated it farther from the creation of the payload. This
> doesn't look like a hot code path, and if this assumption is ever broken
> then getting cute with this won't have been it.
> ------------------------------
>
> *pkg/sql/jobs/jobs.go, line 444 at r12
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEw2ENEJblhytb18U:-KqQEw2ENEJblhytb18V:b7hxi8d>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L444>):*
>
> 	return j.update(ctx, StatusRunning, func(payload *Payload) (bool, error) {
> 		if !payload.Lease.Equal(oldLease) {
> 			return false, errors.New("existing lease did not match expected lease")
>
> Mind including the leases in the error?
> ------------------------------
>
> *pkg/sql/jobs/jobs.go, line 455 at r12
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFC3_XBjTb1E-VUkA:-KqQFC3_XBjTb1E-VUkB:b-p21vfo>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L455>):*
>
> func (j *Job) doCancel() {
> 	j.cancel()
>
> Job has a cancel method and a doCancel method, neither of which takes any
> parameters? That seems likely to cause confusion.
> ------------------------------
>
> *pkg/sql/jobs/jobs.proto, line 33 at r12
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFKoCjgC8lQRHnVOE:-KqQFKoCjgC8lQRHnVOF:bso2qmb>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.proto#L33>):*
>
>   ];
>   // The epoch of the lease holder's node liveness entry.
>   int64 epoch = 2;
>
> Piggy-backing on node liveness is nice from a reuse perspective, but it
> seems like the wrong model. Our epoch-based leases are designed to time out
> quickly, in order to minimize unavailability to transactional workloads.
> "Quickly" means 9 seconds today, but if Spencer has his way it'll soon be
> 4.5 seconds (#16164
> <https://github.com/cockroachdb/cockroach/issues/16164>). This means that
> anything that disrupts the coordinator node or the node liveness range's
> leaseholder for a couple seconds will cancel the job.
>
> How expensive is it for a job to get canceled? If it's a fairly big deal,
> then I'd argue we need to be using longer leases for them to provide extra
> cushion.
>
> For backward-compatible migrations, we're using a duration of 1 minute (
> https://github.com/cockroachdb/cockroach/blob/master/pkg/migration/sqlmigrations/migrations.go#L39),
> and our migrations take much less time than these jobs. The lease
> implementation that they use (
> https://github.com/cockroachdb/cockroach/blob/master/pkg/internal/client/lease.go)
> could be adapted for this if needed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqQHS4JiwGMa-gSHC9h:b-wvofux>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17301#issuecomment-319237274>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IB3Bk4ADfrJH1rkyQKIBvxc3oCa0ks5sTneNgaJpZM4OnKQy>
> .
>

benesch commented:
Review status: 0 of 15 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 693 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqPeinC_EJP5CQJXOkW:-KqT6_FuRULVOQ_wYwr5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c5a2a59bb58706965b695913d70693ccaacac0a/pkg/ccl/sqlccl/backup_test.go#L693)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This should probably be two tests. (or at least two subtests)
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.go, line 254 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQDvJ9SmFhbaK6E8hO:-KqTC4z8e_f7hzBDqB8Y:b-d5djv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L254)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Would there be any harm in including the error in this message?
</blockquote></details>

It's potentially an irrelevant error if the job fails the instant before it's canceled—and if the job is canceled before it fails, the error will always be "context canceled." Seems like it be confusing to include, plus it doesn't give us much value even when it is right.

---

*[pkg/sql/jobs/jobs.go, line 330 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEEbbPFpWTFJCU9fd:-KqTBiVCnubwQqY813bx:btcse5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L330)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Is there any reason not to lock though? I know you didn't write this, but you've now separated it farther from the creation of the payload. This doesn't look like a hot code path, and if this assumption is ever broken then getting cute with this won't have been it.
</blockquote></details>

Code path isn't even a little bit hot. Done.

---

*[pkg/sql/jobs/jobs.go, line 444 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQEw2ENEJblhytb18U:-KqTBV8Dgc6IoU2e3BR8:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L444)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Mind including the leases in the error?
</blockquote></details>

Good idea. Done.

---

*[pkg/sql/jobs/jobs.go, line 455 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFC3_XBjTb1E-VUkA:-KqTBWHx25c-yrbr-vZD:b-1hdvzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.go#L455)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

`Job` has a `cancel` method and a `doCancel` method, neither of which takes any parameters? That seems likely to cause confusion.
</blockquote></details>

Ok, switched to `cancel()` and `cancelFn`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
If we don't lose much work upon cancellation, then I'm perfectly happy with this as is. Nice work on making it easy to resume jobs!

---

Review status: 0 of 15 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqTR1xcQwO3RYNVmx_i:b-738yby)*
<!-- Sent from Reviewable.io -->

a-robinson commented on pkg/sql/jobs/registry.go:
```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// TODO(benesch): cancel only jobs from previous epochs. This logic is correct
+	// but too aggressive, as it might cancel jobs created immediately after a
```
It also might cancel jobs even if the node's liveness never expired. For example:

1. The node sends a heartbeat to update its liveness expiration
1. The heartbeat applies before the node's liveness expires
1. The response to the heartbeat takes a while to get back to the node, so its locally cached liveness is a little out of date and indicates that the node is no longer live.

In these cases, we will cancel things here but we don't actually need to. This can also be left as a TODO, but it's another thing that could be improved.

a-robinson commented on pkg/sql/jobs/registry.go:
```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
+		}
+
+		if payload.Lease == nil {
+			// If the lease is missing, it simply means the job does not yet support
```
Should you do anything here to "cancel" jobs that are in the running state but won't ever finish? Or is there some other way for old abandoned jobs to get cleaned up?

a-robinson commented on pkg/sql/jobs/registry.go:
```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
+			}
+
+			go func() {
+				err := resumeFn(ctx, &job)
```
It seems worth logging that we're taking over a job.

a-robinson commented on pkg/sql/jobs/registry.go:
```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
+
+// Start polls the current node for liveness failures and cancels all registered
+// jobs if it observes a failure.
+func (r *Registry) Start(
```
I'd recommend getting the node's epoch once before waiting out the `cancelInterval`, which I guess would mean just running `maybeCancelJobs` once up front.

benesch commented:
:tada: 

---

Review status: 0 of 15 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/jobs.proto, line 33 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqQFKoCjgC8lQRHnVOE:-KqTqnJ3pnn_IQJ7xcLu:b-cubbio) ([raw file](https://github.com/cockroachdb/cockroach/blob/137973e7bd9f906653b0f0d80f58b7bc1ce5ea46/pkg/sql/jobs/jobs.proto#L33)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Piggy-backing on node liveness is nice from a reuse perspective, but it seems like the wrong model. Our epoch-based leases are designed to time out quickly, in order to minimize unavailability to transactional workloads. "Quickly" means 9 seconds  today, but if Spencer has his way it'll soon be 4.5 seconds (#16164). This means that anything that disrupts the coordinator node or the node liveness range's leaseholder for a couple seconds will cancel the job.

How expensive is it for a job to get canceled? If it's a fairly big deal, then I'd argue we need to be using longer leases for them to provide extra cushion.

For backward-compatible migrations, we're using a duration of 1 minute (https://github.com/cockroachdb/cockroach/blob/master/pkg/migration/sqlmigrations/migrations.go#L39), and our migrations take much less time than these jobs. The lease implementation that they use (https://github.com/cockroachdb/cockroach/blob/master/pkg/internal/client/lease.go) could be adapted for this if needed.
</blockquote></details>

(Discussion about this resolved in the main thread.)

---

*[pkg/sql/jobs/registry.go, line 111 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-111:-KqToV_eembyDNl-gvHF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L111)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'd recommend getting the node's epoch once before waiting out the `cancelInterval`, which I guess would mean just running `maybeCancelJobs` once up front.
</blockquote></details>

Done.

---

*[pkg/sql/jobs/registry.go, line 154 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-154:-KqTpb1YHrkgs-sMURWm:b-g5tir) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L154)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It also might cancel jobs even if the node's liveness never expired. For example:

1. The node sends a heartbeat to update its liveness expiration
1. The heartbeat applies before the node's liveness expires
1. The response to the heartbeat takes a while to get back to the node, so its locally cached liveness is a little out of date and indicates that the node is no longer live.

In these cases, we will cancel things here but we don't actually need to. This can also be left as a TODO, but it's another thing that could be improved.
</blockquote></details>

Ah, good point. I've updated the TODO here with the scenario you describe.

---

*[pkg/sql/jobs/registry.go, line 210 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-210:-KqTnC04xk86CKwRiWBh:b-75qefa) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L210)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should you do anything here to "cancel" jobs that are in the running state but won't ever finish? Or is there some other way for old abandoned jobs to get cleaned up?
</blockquote></details>

Yes, we definitely should. Even in 1.0 we would leak old abandoned jobs. Unfortunately there's no easy way to distinguish between a job currently being properly run by a v1.0 node and a job that's been abandoned by its v1.0 coordinator: they'll both have nil leases. 

The solution is probably to set the status of all jobs over _x_ days old as `abandoned`, but how do you feel about leaving that to another PR?

---

*[pkg/sql/jobs/registry.go, line 244 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-244:-KqTnNSkt1YDM21p2jPg:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L244)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It seems worth logging that we're taking over a job.
</blockquote></details>

Good call, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqTqlnFMVQ-lHbckU6n:bjwa7hh)*
<!-- Sent from Reviewable.io -->

danhhz commented:
I have a few more style/cleanliness issues as well as bikeshedding on the logging, but it's not worth all the rebasing. I'll take one more pass over what you merge and we can do it as a followup. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 18 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/registry.go, line 260 at r20](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqTmnTYnHstwKW6_fws:-KqTmnTYnHstwKW6_fwt:b-f0mb1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c819e8a8d966fa26050fa21939164306c9d4635/pkg/sql/jobs/registry.go#L260)):*
> ```Go
> 	r.mu.AssertHeld()
> 	for jobID, job := range r.mu.jobs {
> 		if log.V(2) {
> ```

this is important enough information that it warrants being logged unconditionally

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqUEqqskFWMt4qoO7yN:b-y54qet)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 18 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/registry.go, line 260 at r20](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqTmnTYnHstwKW6_fws:-KqUG4xNQCvRzFIf-L1K:blwjt5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c819e8a8d966fa26050fa21939164306c9d4635/pkg/sql/jobs/registry.go#L260)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this is important enough information that it warrants being logged unconditionally
</blockquote></details>

whoops. forgot this was here. an example of what i meant for followup

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Registry code LGTM

---

Reviewed 1 of 15 files at r25.
Review status: 1 of 18 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/jobs/registry.go, line 154 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-154:-KqUIYcCT6Hizn_5lDi4:b-sdr368) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L154)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ah, good point. I've updated the TODO here with the scenario you describe.
</blockquote></details>

Thanks. What you wrote isn't quite correct, though. It's not about when the job was created, it's about falsely determining that our node liveness expired while the job was running. The potential reasons for it to fail would be the time when this function called `nl.Self()` to get the liveness object and what `r.clock.Now()` returns below. You could imagine this happening in a race-enabled build even without network delay if there was enough delay between when `nl.Self()` returned above and `r.clock.Now()` gets called below.

---

*[pkg/sql/jobs/registry.go, line 210 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-KqAQPgsnzjcNDbwWTow-r24-210:-KqUJAYX668hId7k0DVb:b-6bje2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/af72d3f8ef682fd61ac17be6bf301177bbe26ca1/pkg/sql/jobs/registry.go#L210)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Yes, we definitely should. Even in 1.0 we would leak old abandoned jobs. Unfortunately there's no easy way to distinguish between a job currently being properly run by a v1.0 node and a job that's been abandoned by its v1.0 coordinator: they'll both have nil leases. 

The solution is probably to set the status of all jobs over _x_ days old as `abandoned`, but how do you feel about leaving that to another PR?
</blockquote></details>

Yeah, I don't mean that you should do it now, I was just making sure you had thought through it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17301#-:-KqUJaSRjQmI_V0GEOlQ:b-fj453)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ok, some correctness results are in. @danhhz ran a restore on a lapis, using a58e3c0, under serious chaos. The first half of the restore took a node out for ~10s every 1m; that proved too disruptive to node liveness, so we switched to taking a node out every 2m for the second half of the restore. The last five minutes were run without chaos and without node 1, which was manually killed five minutes before the restore finished. 

In the interest of expedience we only checked the experimental fingerprints of `tpch.customer`, but it matched the results of a restore without chaos using the latest master! See below.

```root@localhost:26257/> show experimental_fingerprints from table tpch_golden.customer;
+---------+----------------------+
|  Index  |     Fingerprint      |
+---------+----------------------+
| primary | -5755662949776301116 |
| c_ck    | -7901840833801052955 |
| c_nk    | -7920475550559869418 |
+---------+----------------------+
(3 rows)
root@localhost:26257/> show experimental_fingerprints from table tpch.customer;
+---------+----------------------+
|  Index  |     Fingerprint      |
+---------+----------------------+
| primary | -5755662949776301116 |
| c_ck    | -7901840833801052955 |
| c_nk    | -7920475550559869418 |
+---------+----------------------+
(3 rows)```

Anyway, I think I've addressed everything! @a-robinson, could you take one last look at that comment?

a-robinson reviewed (approved):
I don't think you'd have to block ongoing work while waiting to see if a heartbeat was successful, but good enough.

benesch commented:
TFTRs—especially on such short notice, @a-robinson! I owe you one.

danhhz commented:
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉
